File: backend/services/llm.service.js
/**
 * Multilingual AI Briefings Generator
 * Uses ONLY canonical telemetry data — never fabricates values.
 *
 * STRICT RULE:
 * The deterministic backend calculates overall monitoring status (CRITICAL, HIGH, ELEVATED, NORMAL).
 * The LLM MUST NOT calculate or override this status; it only generates natural-language
 * explanations grounded in the already-calculated status.
 */

/**
 * Extract a numeric value from either a raw number or a structured {value, unit, ...} object.
 */
function extractValue(field) {
  if (field === null || field === undefined) return null;
  if (typeof field === "object" && "value" in field) {
    return field.value !== null && field.value !== undefined ? Number(field.value) : null;
  }
  const n = Number(field);
  return Number.isFinite(n) ? n : null;
}

function safeNum(val, fallbackLabel = "Unavailable") {
  if (val === null || val === undefined || !Number.isFinite(val)) return fallbackLabel;
  return val.toFixed(1);
}

/**
 * Generate risk-appropriate recommendations based on validated conditions.
 */
function generateRecommendations({ riskClass, overallStatus, cwcStatus, basinName, exposedPopStr, language }) {
  const actions = {
    LOW: {
      en: [
        `Continue normal hydrological monitoring for ${basinName}.`,
        `No emergency resource deployment required at this time.`,
        `Review forecast updates every 12 hours for potential changes.`
      ],
      hi: [
        `${basinName} के लिए सामान्य जल-विज्ञान निगरानी जारी रखें।`,
        `इस समय किसी आपातकालीन संसाधन तैनाती की आवश्यकता नहीं है।`,
        `हर 12 घंटे में पूर्वानुमान अपडेट की समीक्षा करें।`
      ]
    },
    MODERATE: {
      en: [
        `Increase monitoring frequency for rainfall and river levels in ${basinName}.`,
        `Prepare local emergency resources for potential deployment.`,
        `Alert district-level disaster management authorities.`,
        `Ensure communication channels with at-risk communities are active.`
      ],
      hi: [
        `${basinName} में वर्षा और नदी जलस्तर की निगरानी बढ़ाएं।`,
        `स्थानीय आपातकालीन संसाधनों को तैनाती के लिए तैयार रखें।`,
        `जिला-स्तरीय आपदा प्रबंधन अधिकारियों को सतर्क करें।`
      ]
    },
    HIGH: {
      en: [
        `Pre-position State Disaster Response Force (SDRF) boats along low-lying embankments in ${basinName}.`,
        exposedPopStr !== "Unavailable"
          ? `Issue early SMS advisories to ${exposedPopStr} residents living within the active floodplain.`
          : `Prepare targeted public advisories for identified risk areas once exposure data is available.`,
        `Inspect structural integrity of river sluice gates, bridges, and culverts given precipitation surge.`,
        `Deploy emergency medical and drinking water relief kits to designated relief camps.`
      ],
      hi: [
        `${basinName} के निचले तटबंधों पर SDRF को तैनात करें।`,
        exposedPopStr !== "Unavailable"
          ? `बाढ़ क्षेत्र में रहने वाले ${exposedPopStr} नागरिकों को एसएमएस चेतावनी जारी करें।`
          : `जोखिम क्षेत्रों के लिए लक्षित सार्वजनिक सलाह तैयार करें।`,
        `पुलों, तटबंधों और जल निकासी द्वारों का निरीक्षण करें।`,
        `राहत शिविरों में पेयजल और प्राथमिक चिकित्सा किट सुरक्षित करें।`
      ]
    },
    SEVERE: {
      en: [
        `IMMEDIATE: Activate full emergency response coordination for ${basinName}.`,
        exposedPopStr !== "Unavailable"
          ? `Evacuate ${exposedPopStr} residents from active floodplain zones.`
          : `Begin evacuation of identified flood-risk areas immediately.`,
        `Deploy all available NDRF/SDRF resources to critical embankments.`,
        `Establish emergency communication links with all block-level officers.`,
        `Prepare helicopter/boat rescue for stranded populations.`
      ],
      hi: [
        `तत्काल: ${basinName} के लिए पूर्ण आपातकालीन प्रतिक्रिया सक्रिय करें।`,
        exposedPopStr !== "Unavailable"
          ? `बाढ़ क्षेत्र से ${exposedPopStr} नागरिकों को तुरंत स्थानांतरित करें।`
          : `पहचाने गए बाढ़ जोखिम क्षेत्रों से तुरंत निकासी शुरू करें।`,
        `सभी उपलब्ध NDRF/SDRF संसाधनों को तैनात करें।`,
        `हेलीकॉप्टर/नाव बचाव की तैयारी करें।`
      ]
    }
  };

  // Base actions on overall monitoring status if higher than risk class
  const effectiveLevel = (overallStatus === "CRITICAL" || overallStatus === "SEVERE") ? "SEVERE"
    : overallStatus === "HIGH" ? "HIGH"
    : overallStatus === "ELEVATED" ? "MODERATE"
    : (actions[riskClass] ? riskClass : "LOW");

  const level = actions[effectiveLevel] || actions.LOW;
  const langActions = level[language] || level.en;

  // If CWC gauge is above warning or danger but ML says LOW, add an explicit independence note
  if ((cwcStatus === "ABOVE_WARNING" || cwcStatus === "ABOVE_DANGER" || cwcStatus === "EXTREME") && riskClass === "LOW") {
    const caveat = language === "hi"
      ? `⚠️ आधिकारिक CWC अवलोकन: नदी जलस्तर चेतावनी/खतरे के स्तर से ऊपर है। यद्यपि AI मॉडल कम जलभराव का अनुमान लगाता है, जल-स्तर की स्थिति के कारण बढ़ी हुई निगरानी आवश्यक है।`
      : `⚠️ Official CWC Ground Truth: River stage is ${cwcStatus.replace("_", " ")}. Although the AI model currently estimates low inundation probability, observed river conditions dictate ${overallStatus} monitoring.`;
    return [caveat, ...langActions];
  }

  return langActions;
}

/**
 * Multilingual AI Briefings Generator
 */
export function generateDisasterBriefing({ telemetry, language = "en" }) {
  const basinName =
    telemetry?.location?.basin_name ||
    telemetry?.location?.district ||
    "Regional Basin";

  const probability = extractValue(telemetry?.prediction?.flood_probability_pct);
  const riskClass = telemetry?.prediction?.risk_class || "UNKNOWN";

  const rain72h = extractValue(telemetry?.current_weather?.rainfall_72h);
  const rain24h = extractValue(telemetry?.current_weather?.rainfall_24h);

  const observedHydro = telemetry?.cwc_ground_truth || telemetry?.observed_hydrology_status || {};
  const overall = telemetry?.overall_monitoring || {};
  const overallStatus = overall.status || (riskClass === "HIGH" || riskClass === "SEVERE" || riskClass === "VERY HIGH" ? "HIGH ALERT" : riskClass === "MODERATE" || riskClass === "MEDIUM" ? "WATCH" : "NORMAL");

  const riverStageVal = observedHydro.water_level_m ?? observedHydro.stageM ?? telemetry?.hydrology?.river_stage?.value ?? null;
  const cwcStatus = observedHydro.condition || observedHydro.status || "UNAVAILABLE";

  const exposedPopVal = extractValue(telemetry?.exposure?.population);
  const exposedPopStr = exposedPopVal !== null ? exposedPopVal.toLocaleString() : "Unavailable";

  const probStr = safeNum(probability, "Unavailable");
  const rain72hStr = safeNum(rain72h, "Unavailable");

  const recommendations = generateRecommendations({
    riskClass,
    overallStatus,
    cwcStatus,
    basinName,
    exposedPopStr,
    language
  });

  const riverStageDesc = (riverStageVal !== null && riverStageVal !== undefined && Number.isFinite(Number(riverStageVal)))
    ? `${Number(riverStageVal).toFixed(2)} m (${String(cwcStatus).replace("_", " ")})`
    : `Unavailable (${observedHydro.reason || observedHydro.failureReason || "Live CWC API unavailable"})`;

  const contentMap = {
    en: {
      headline: overall.message
        ? `${overallStatus} Monitoring: ${basinName}`
        : `${riskClass} Flood Risk Detected for ${basinName}`,
      summary: `AI Flood Risk: ${probStr}% (${riskClass}). Observed River Condition: ${riverStageDesc}. Overall Monitoring: ${overallStatus}. ${overall.message || ""}`,
      urgency: overallStatus === "CRITICAL"
        ? "CRITICAL FLOOD DEFENSE EMERGENCY"
        : overallStatus === "HIGH"
          ? "FLOOD ALERT / HIGH HAZARD"
          : overallStatus === "ELEVATED"
            ? "ELEVATED CATCHMENT SURVEILLANCE"
            : "NORMAL MONITORING STATUS"
    },
    hi: {
      headline: `${basinName}: ${overallStatus === "CRITICAL" ? "अत्यधिक आपातकालीन" : overallStatus === "HIGH" ? "उच्च सतर्कता" : overallStatus === "ELEVATED" ? "उन्नत" : "सामान्य"} निगरानी स्थिति`,
      summary: `AI जोखिम: ${probStr}% (${riskClass})। प्रेक्षित नदी जलस्तर: ${riverStageDesc}। समग्र निगरानी: ${overallStatus}। AI पूर्वानुमान और आधिकारिक CWC प्रेक्षण स्वतंत्र संकेत हैं।`,
      urgency: overallStatus === "CRITICAL" ? "अत्यधिक आपातकालीन चेतावनी" : overallStatus === "HIGH" ? "तत्काल राहत और निकासी सतर्कता" : overallStatus === "ELEVATED" ? "सक्रिय जलसंभर निगरानी" : "सामान्य निगरानी"
    }
  };

  const selected = contentMap[language] || contentMap.en;

  return {
    language,
    headline: selected.headline,
    summary: selected.summary,
    urgency: selected.urgency,
    actions: recommendations,
    ai_provenance: "AI prediction and observed river conditions are independent signals. Overall monitoring status calculated deterministically by ChetakAI decision engine.",
    generated_at: new Date().toISOString()
  };
}

/**
 * Interactive Copilot Chat responder
 */
export async function chatWithCopilot({ message, telemetry, language = "en", history = [] }) {
  const basinName = telemetry?.location?.basin_name || telemetry?.location?.district || "Target Basin";
  const probability = extractValue(telemetry?.prediction?.flood_probability_pct);
  const riskClass = telemetry?.prediction?.risk_class || "UNKNOWN";
  const overallStatus = telemetry?.overall_monitoring?.status || "NORMAL";
  const overallMessage = telemetry?.overall_monitoring?.message || "";
  const observedHydro = telemetry?.observed_hydrology_status || {};
  const riverStageVal = observedHydro.stageM;
  const cwcStatus = observedHydro.status || "UNAVAILABLE";

  const rain24h = extractValue(telemetry?.current_weather?.rainfall_24h);
  const rain72h = extractValue(telemetry?.current_weather?.rainfall_72h);
  const slope = extractValue(telemetry?.terrain?.mean_slope_deg);
  const soilRunoff = extractValue(telemetry?.soil?.soil_runoff_proxy);
  const exposedPop = extractValue(telemetry?.exposure?.population);
  const exposedPopStr = exposedPop !== null ? exposedPop.toLocaleString() : "Unavailable";

  const probStr = safeNum(probability, "Unavailable");
  const rain72hStr = safeNum(rain72h, "Unavailable");
  const rain24hStr = safeNum(rain24h, "Unavailable");
  const slopeStr = safeNum(slope, "Unavailable");
  const soilStr = soilRunoff !== null ? String(soilRunoff) : "Unavailable";

  const cwcDesc = riverStageVal !== null
    ? `${riverStageVal.toFixed(2)} m (${cwcStatus})`
    : `Unavailable (${observedHydro.failureReason || "Live API unavailable"})`;

  // Check if external GEMINI_API_KEY is configured
  const geminiKey = process.env.GEMINI_API_KEY;
  if (geminiKey) {
    try {
      const response = await fetch(
        `https://generativelanguage.googleapis.com/v1beta/models/gemini-2.0-flash:generateContent?key=${geminiKey}`,
        {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({
            contents: [
              {
                role: "user",
                parts: [
                  {
                    text: `You are ChetakAI Flood Intelligence Copilot, an expert hydrologist and disaster defense AI.
Ground truth telemetry and dual-signal monitoring for the user's active coordinate:
- Basin: ${basinName}
- AI ML Inundation Probability: ${probStr}% (${riskClass} Risk)
- Observed CWC River Stage: ${cwcDesc}
- Overall Monitoring Status: ${overallStatus} (Pre-calculated deterministically)
- Note: AI prediction and observed river conditions are independent signals.
- 24h Rainfall: ${rain24hStr} mm, 72h Cumulative: ${rain72hStr} mm
- Topographical Slope: ${slopeStr} degrees
- Soil Runoff Index: ${soilStr}
- Exposed Population: ${exposedPopStr}

User Question: "${message}"
Respond in language: ${language}.
Provide an authoritative, clear, actionable answer grounded in the above physical facts. Never fabricate numbers or override the pre-calculated overall monitoring status. If a value is "Unavailable", explain why.`
                  }
                ]
              }
            ]
          })
        }
      );

      if (response.ok) {
        const json = await response.json();
        const reply = json?.candidates?.[0]?.content?.parts?.[0]?.text;
        if (reply) {
          return {
            reply: reply.trim(),
            source: "Gemini 2.0 Realtime Physical Grounding",
            language
          };
        }
      }
    } catch (err) {
      console.warn("External Gemini API call skipped:", err.message);
    }
  }

  // Resilient Local Intelligence Engine
  const q = message.toLowerCase();
  let reply = "";
  const localized = {
    en: {
      cwc: `CWC river stage is ${cwcDesc}. Overall monitoring status is ${overallStatus}.`,
      risk: `${overallMessage} AI flood risk is ${probStr}%. Overall monitoring status is ${overallStatus}.`,
      evacuate: `Evacuation planning should target identified flood-risk zones. Current status: ${overallStatus}.`,
      rain: `Current rainfall is ${rain24hStr} mm in 24 hours and ${rain72hStr} mm over 72 hours.`,
      summary: `For ${basinName}, AI flood probability is ${probStr}% (${riskClass}) and overall monitoring status is ${overallStatus}.`
    },
    hi: {
      cwc: `CWC नदी जलस्तर ${cwcDesc} है। समग्र निगरानी स्थिति ${overallStatus} है।`,
      risk: `${overallMessage} एआई बाढ़ जोखिम ${probStr}% है। समग्र निगरानी स्थिति ${overallStatus} है।`,
      evacuate: `निकासी योजना में पहचाने गए बाढ़ जोखिम क्षेत्रों को प्राथमिकता दें। वर्तमान स्थिति: ${overallStatus}।`,
      rain: `पिछले 24 घंटों में वर्षा ${rain24hStr} मिमी और 72 घंटों में ${rain72hStr} मिमी है।`,
      summary: `${basinName} के लिए एआई बाढ़ संभावना ${probStr}% (${riskClass}) है और समग्र निगरानी स्थिति ${overallStatus} है।`
    },
    bn: {
      cwc: `CWC নদীর জলস্তর ${cwcDesc}। সামগ্রিক পর্যবেক্ষণ অবস্থা ${overallStatus}।`,
      risk: `${overallMessage} এআই বন্যার ঝুঁকি ${probStr}%। সামগ্রিক পর্যবেক্ষণ অবস্থা ${overallStatus}।`,
      evacuate: `চিহ্নিত বন্যা ঝুঁকিপূর্ণ এলাকায় উচ্ছেদ পরিকল্পনা অগ্রাধিকার দিন। বর্তমান অবস্থা: ${overallStatus}।`,
      rain: `গত ২৪ ঘণ্টায় বৃষ্টিপাত ${rain24hStr} মিমি এবং ৭২ ঘণ্টায় ${rain72hStr} মিমি।`,
      summary: `${basinName}-এর এআই বন্যার সম্ভাবনা ${probStr}% (${riskClass}) এবং সামগ্রিক পর্যবেক্ষণ অবস্থা ${overallStatus}।`
    },
    mr: {
      cwc: `CWC नदीची पातळी ${cwcDesc} आहे. एकूण निरीक्षण स्थिती ${overallStatus} आहे.`,
      risk: `${overallMessage} एआय पूर जोखीम ${probStr}% आहे. एकूण निरीक्षण स्थिती ${overallStatus} आहे.`,
      evacuate: `ओळखलेल्या पूर जोखीम क्षेत्रांना स्थलांतर नियोजनात प्राधान्य द्या. सध्याची स्थिती: ${overallStatus}.`,
      rain: `मागील 24 तासांत पाऊस ${rain24hStr} मिमी आणि 72 तासांत ${rain72hStr} मिमी आहे.`,
      summary: `${basinName} साठी एआय पूर संभाव्यता ${probStr}% (${riskClass}) आणि एकूण निरीक्षण स्थिती ${overallStatus} आहे.`
    },
    te: {
      cwc: `CWC నది నీటి మట్టం ${cwcDesc}. మొత్తం పర్యవేక్షణ స్థితి ${overallStatus}.`,
      risk: `${overallMessage} AI వరద ప్రమాదం ${probStr}%. మొత్తం పర్యవేక్షణ స్థితి ${overallStatus}.`,
      evacuate: `గుర్తించిన వరద ప్రమాద ప్రాంతాలకు తరలింపు ప్రణాళికలో ప్రాధాన్యత ఇవ్వండి. ప్రస్తుత స్థితి: ${overallStatus}.`,
      rain: `గత 24 గంటల్లో వర్షపాతం ${rain24hStr} మిమీ, 72 గంటల్లో ${rain72hStr} మిమీ.`,
      summary: `${basinName} కోసం AI వరద సంభావ్యత ${probStr}% (${riskClass}), మొత్తం పర్యవేక్షణ స్థితి ${overallStatus}.`
    },
    ta: {
      cwc: `CWC ஆற்றின் நீர்மட்டம் ${cwcDesc}. ஒட்டுமொத்த கண்காணிப்பு நிலை ${overallStatus}.`,
      risk: `${overallMessage} AI வெள்ள அபாயம் ${probStr}%. ஒட்டுமொத்த கண்காணிப்பு நிலை ${overallStatus}.`,
      evacuate: `அடையாளம் காணப்பட்ட வெள்ள அபாயப் பகுதிகளுக்கு வெளியேற்றத் திட்டத்தில் முன்னுரிமை அளிக்கவும். தற்போதைய நிலை: ${overallStatus}.`,
      rain: `கடந்த 24 மணி நேர மழை ${rain24hStr} மிமீ, 72 மணி நேர மழை ${rain72hStr} மிமீ.`,
      summary: `${basinName}க்கான AI வெள்ள நிகழ்தகவு ${probStr}% (${riskClass}), ஒட்டுமொத்த கண்காணிப்பு நிலை ${overallStatus}.`
    }
  }[language] || null;

  if (q.includes("cwc") || q.includes("stage") || q.includes("river") || q.includes("level")) {
    reply = localized?.cwc || `CWC River Stage telemetry: ${cwcDesc}. Overall monitoring status is ${overallStatus}.`;
  } else if (q.includes("why") && (q.includes("risk") || q.includes("status") || q.includes("monitoring") || q.includes("elevated"))) {
    reply = localized?.risk || `${overallMessage} AI Flood Risk: ${probStr}%. Overall monitoring status is ${overallStatus}.`;
  } else if (q.includes("evacuat") || q.includes("safe") || q.includes("shelter")) {
    reply = localized?.evacuate || `Evacuation planning should target identified flood-risk zones. Current status: ${overallStatus}.`;
  } else if (q.includes("rain") || q.includes("weather")) {
    reply = localized?.rain || `Current rainfall is ${rain24hStr} mm in 24 hours and ${rain72hStr} mm over 72 hours.`;
  } else {
    reply = localized?.summary || `Based on current telemetry for ${basinName}, AI flood probability is ${probStr}% (${riskClass}) and overall monitoring status is ${overallStatus}.`;
  }

  return {
    reply,
    source: "ChetakAI Grounded Physics Engine",
    language
  };
}

File: backend/app/services/llm_service.py
from typing import Any, Dict, List, Optional


def extract_val(field: Any) -> Optional[float]:
    if field is None:
        return None
    if isinstance(field, dict) and "value" in field:
        val = field["value"]
        if val is not None:
            try:
                return float(val)
            except (ValueError, TypeError):
                return None
    try:
        return float(field)
    except (ValueError, TypeError):
        return None


def generate_disaster_briefing(telemetry: Dict[str, Any], language: str = "en") -> Dict[str, Any]:
    """
    Generates a deterministic multilingual disaster briefing grounded in
    the validated multi-signal telemetry.
    """
    loc = telemetry.get("location", {})
    pred = telemetry.get("prediction", {})
    alert = telemetry.get("alert", {})
    overall = telemetry.get("overall_monitoring", {})
    cwc = telemetry.get("cwc_ground_truth", {}) or {}
    weather = telemetry.get("current_weather", {})
    fallback = telemetry.get("fallback_environmental", {})

    basin_name = loc.get("basin_name") or loc.get("city") or loc.get("district") or "the monitored basin"
    overall_status = overall.get("status") or alert.get("level") or "NORMAL"
    ai_risk = pred.get("risk_class") or "LOW"
    ai_prob_pct = pred.get("flood_probability_pct") or 0.0

    r24 = extract_val(weather.get("rainfall_24h")) or fallback.get("rainfall_mm") or 0.0
    cwc_status = cwc.get("status") or "UNAVAILABLE"
    cwc_stage = cwc.get("water_level_m")
    cwc_danger = cwc.get("danger_level_m")

    # Multilingual headlines and summaries
    if overall_status == "CRITICAL":
        headline_en = f"CRITICAL FLOOD EMERGENCY — {basin_name.upper()}"
        headline_hi = f"गंभीर बाढ़ आपातकाल — {basin_name.upper()}"
        summary_en = f"Active severe flood alert. CWC gauge confirms elevated stage exceeding thresholds with AI inundation probability of {ai_prob_pct:.1f}%. Immediate evacuation and emergency protocols active."
        summary_hi = f"सक्रिय गंभीर बाढ़ चेतावनी। सीडब्ल्यूसी गेज खतरनाक जलस्तर की पुष्टि करता है। कृत्रिम बुद्धिमत्ता मॉडल {ai_prob_pct:.1f}% बाढ़ संभावना दिखाता है। तत्काल निकासी आवश्यक है।"
        actions = [
            "Evacuate vulnerable riverbank and low-lying zones immediately.",
            "Activate emergency shelters and deploy National Disaster Response Force (NDRF) teams.",
            "Cut power to inundated substations and monitor flood retaining bunds continuously."
        ]
    elif overall_status == "HIGH ALERT":
        headline_en = f"HIGH FLOOD ALERT DEPLOYED — {basin_name.upper()}"
        headline_hi = f"उच्च बाढ़ चेतावनी जारी — {basin_name.upper()}"
        summary_en = f"High alert condition active. Elevated precipitation ({r24:.1f} mm/24h) and physical hydrological loading detected. River telemetry indicates rapid rise."
        summary_hi = f"उच्च चेतावनी स्थिति सक्रिय। पिछले 24 घंटों में {r24:.1f} मिमी भारी वर्षा और जलग्रहण क्षेत्र में बढ़ा हुआ जलप्रवाह दर्ज किया गया है।"
        actions = [
            "Alert district disaster response teams and stage rescue boats.",
            "Notify riverbank communities to avoid low-lying crossings.",
            "Verify backup communication systems and inspect flood embankments."
        ]
    elif overall_status == "WATCH":
        headline_en = f"FLOOD WATCH ACTIVE — {basin_name.upper()}"
        headline_hi = f"बाढ़ निगरानी सक्रिय — {basin_name.upper()}"
        summary_en = f"Hydrological flood watch active for {basin_name}. Environmental indicators and cumulative rainfall warrant continuous monitoring."
        summary_hi = f"{basin_name} के लिए बाढ़ निगरानी सक्रिय है। वर्षा और पर्यावरणीय संकेतकों पर लगातार नजर रखी जा रही है।"
        actions = [
            "Monitor hourly rainfall and upstream reservoir discharge reports.",
            "Ensure emergency equipment and communication channels are operational.",
            "Advise local administrations to prepare emergency shelters if precipitation accelerates."
        ]
    else:
        headline_en = f"NORMAL HYDROLOGICAL MONITORING — {basin_name.upper()}"
        headline_hi = f"सामान्य जल-विज्ञान निगरानी — {basin_name.upper()}"
        summary_en = f"Hydrological conditions remain within baseline parameters across {basin_name}. AI probability is {ai_prob_pct:.1f}% with normal catchment loading."
        summary_hi = f"{basin_name} में जल-विज्ञान की स्थिति सामान्य सीमा के भीतर है। बाढ़ की संभावना {ai_prob_pct:.1f}% है।"
        actions = [
            f"Continue routine hydrological monitoring across {basin_name}.",
            "No emergency deployment or evacuation required at this time.",
            "Review meteorological forecast updates every 12 hours."
        ]

    is_hindi = language == "hi"

    return {
        "headline": headline_hi if is_hindi else headline_en,
        "summary": summary_hi if is_hindi else summary_en,
        "key_risks": [
            f"24-Hour Cumulative Rainfall: {r24:.1f} mm",
            f"CWC Telemetry Status: {cwc_status} ({f'{cwc_stage:.2f} m' if cwc_stage is not None else 'Unavailable'})",
            f"AI Model Probability: {ai_prob_pct:.1f}% ({ai_risk})"
        ],
        "recommendations": actions,
        "language": language
    }


async def chat_with_copilot(
    message: str,
    telemetry: Dict[str, Any],
    language: str = "en",
    history: Optional[List[Dict[str, Any]]] = None
) -> Dict[str, Any]:
    """
    Context-aware Copilot Chat for flood decision support.
    Answers queries grounded strictly in current telemetry.
    """
    loc = telemetry.get("location", {})
    overall = telemetry.get("overall_monitoring", {})
    pred = telemetry.get("prediction", {})
    cwc = telemetry.get("cwc_ground_truth", {})
    weather = telemetry.get("current_weather", {})

    status = overall.get("status", "NORMAL")
    prob = pred.get("flood_probability_pct", 0.0)
    cwc_status = cwc.get("status", "UNAVAILABLE")
    stage = cwc.get("water_level_m")
    city = loc.get("city") or loc.get("district") or loc.get("basin_name") or "monitored location"

    msg_lower = message.lower()
    risk_terms = ("risk", "danger", "status", "जोखिम", "खतरा", "स्थिति", "ঝুঁকি", "धोका", "ప్రమాదం", "ஆபத்து")
    rain_terms = ("rain", "rainfall", "weather", "बारिश", "वर्षा", "मौसम", "বৃষ্টি", "पाऊस", "వర్ష", "மழை")
    river_terms = ("cwc", "gauge", "water level", "river", "नदी", "जलस्तर", "জলস্তর", "नदीची", "నీటి మట్టం", "ஆற்றின்")
    evacuation_terms = ("evacuat", "shelter", "safe", "निकासी", "बचाव", "आश्रय", "উচ্ছেদ", "निवारा", "తరలింపు", "வெளியேற்ற")

    is_hindi = language == "hi"
    is_bengali = language == "bn"
    is_marathi = language == "mr"
    is_telugu = language == "te"
    is_tamil = language == "ta"

    if any(term in msg_lower for term in river_terms):
        if stage is not None:
            if is_hindi:
                reply = f"{city} के निकटतम CWC गेज का वर्तमान जलस्तर {stage:.2f} मीटर है। स्थिति: {cwc.get('condition', 'NORMAL')}।"
            else:
                reply = f"The nearest CWC gauge for {city} is {cwc.get('station_name', 'CWC Gauge')} on River {cwc.get('river', 'Ganga')}. Current observed water level is {stage:.2f} m. Status: {cwc.get('condition', 'NORMAL')}."
        else:
            reply = f"{city} के लिए लाइव CWC गेज स्थिति {cwc_status} है।" if is_hindi else f"For {city}, live CWC gauge telemetry is currently {cwc_status}."
    elif any(term in msg_lower for term in evacuation_terms):
        reply = f"{city} में पहचाने गए बाढ़ जोखिम क्षेत्रों और सुरक्षित आश्रयों को प्राथमिकता दें। वर्तमान स्थिति: {status}." if is_hindi else f"Evacuation planning should target identified flood-risk zones around {city}. Current status: {status}."
    elif any(term in msg_lower for term in risk_terms):
        if is_hindi:
            reply = f"{city} की वर्तमान बाढ़ निगरानी स्थिति {status} है। एआई बाढ़ संभावना {prob:.1f}% है।"
        else:
            reply = f"The current overall flood monitoring status for {city} is {status} ({overall.get('confidence', 'HIGH CONFIDENCE')}). AI flood probability is {prob:.1f}%. Decision basis: {' + '.join(overall.get('basis', ['AI_MODEL']))}."
    elif any(term in msg_lower for term in rain_terms):
        r24 = extract_val(weather.get("rainfall_24h")) or 0.0
        reply = f"{city} में पिछले 24 घंटों की वर्षा {r24:.1f} मिमी है। बेसिन की लगातार निगरानी हो रही है।" if is_hindi else f"Recorded 24-hour rainfall in {city} is {r24:.1f} mm. Catchment conditions are being monitored in real time."
    else:
        reply = f"ChetakAI Monitoring Summary for {city}: Status is {status}. AI inundation probability is {prob:.1f}%, CWC telemetry is {cwc_status}. Operational confidence: {overall.get('confidence', 'MEDIUM CONFIDENCE')}."

    return {
        "reply": reply,
        "status": status,
        "language": language
    }
